src/rollup/types.d.ts

```diff
@@ -466,37 +457,38 @@ export type SequentialPluginHooks =
 	| 'renderChunk'
 	| 'transform';
 
-export type ParallelPluginHooks =
-	| 'banner'
-	| 'buildEnd'
-	| 'buildStart'
-	| 'footer'
-	| 'intro'
-	| 'moduleParsed'
-	| 'outro'
-	| 'renderError'
-	| 'renderStart'
-	| 'writeBundle'
-	| 'closeBundle'
-	| 'closeWatcher'
-	| 'watchChange';
+export type ParallelPluginHooks = Exclude<
+	keyof FunctionPluginHooks | AddonHooks,
+	FirstPluginHooks | SequentialPluginHooks
+>;
 
-interface OutputPluginValueHooks {
-	banner: AddonHook;
-	cacheKey: string;
-	footer: AddonHook;
-	intro: AddonHook;
-	outro: AddonHook;
-}
+export type AddonHooks = 'banner' | 'footer' | 'intro' | 'outro';
 
-export interface Plugin extends Partial<PluginHooks>, Partial<OutputPluginValueHooks> {
-	// for inter-plugin communication
-	api?: any;
+type MakeAsync<T extends (...a: any) => any> = (
+	...a: Parameters<T>
+) => ReturnType<T> | Promise<ReturnType<T>>;
```

Comment:
```suggestion
type MakeAsync<Fn> = Fn extends ({ (this: infer This, ...args: infer Args): infer Return }) ? (
	this: This,
	...args: Args
) => Return | Promise<Return> : never
```

The previous approach loss the `this` type.

https://www.typescriptlang.org/play?#code/C4TwDgpgBAsghgawgQQM4gHYGMCMAeAFSggA9gIMATVKACgDpG4AuKODEASigF4A+NhwE86AKACQjeiygAFOACc4AWwjkFqQn1Hd+UAEpqArgowFwELVAA+chQHtlAS1SXDwE2Yta+AblGioJCwiCjo2ABMeABiGMJQscRkFNR0AN50wAAWLqxOGABmEApQBDmoADRQUooA5qh5hcVQyAr1nI1FJe6eUAC+3AD8YuLZuaXlFRI1bQ0tszq8Aj2mNnaOLm7GpgKsGBAAbsUBQdCJIrRjcwDkWPYY5GTXVXU4rNevz2xtEe91EdddAJrgpthhrqJ8uoCnAsNAAPIAIwAVlA0hIChg3iEkGhMLgYnEpuJMb8cWF8VFYto+gE7hhUMAoGAADZGWr5VhI1EidHiODhLBQTE4WjcPmjcoScQAehlUAAeoNpXKkpAsEyAAb0x7ATVQRFGJm1exM7nS0EeVYgsEQ8R9YkC-HCjARMVo6VXFXypUWsFQG1W8ESWl9IA

Reply:
Great find, completely overlooked this 🙄 Admittedly, typing was the main pain point of this PR.

Reply:
Fixed